### PR TITLE
Add nightly release

### PR DESCRIPTION
**Pull request by kemingy**
- As we discussed in https://github.com/lycheeverse/lychee/pull/1133#issuecomment-1669290206

This action will check if it has changes compared to the last nightly release, if so, it will delete the last nightly release and create a new one.

### CODE REVIEW

**Comment by kemingy** [comment]
According to the GitHub [trigger a workflow from a workflow](https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#triggering-a-workflow-from-a-workflow), the default `GITHUB_TOKEN` cannot trigger another workflow. Thus it requires a personal access token.

You can follow the instructions [here](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#creating-a-fine-grained-personal-access-token) to generate a repository level token and save it to this repo setting with the name `LYCHEE`. It requires the following read/write permission:
- actions
- contents

One thing I'm not sure about is the upload-asset failure: [tested on my fork repo](https://github.com/kemingy/lychee/actions/runs/5840467843/job/15839750545).

File: .github/workflows/nightly.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      tag_name:
+        description: 'Tag name to trigger the release'
+        required: false
+        default: 'nightly'
+
+jobs:
+  trigger:
+    runs-on: ubuntu-latest
+    env:
+      GITHUB_TOKEN: ${{ secrets.LYCHEE }}
```
Comment (mre) [inline .github/workflows/nightly.yml:16]: Can we rename that to `NIGHTLY_BUILD_TOKEN` or something to make the use-case more explicit?

**Comment by mre** [comment]
Thanks for that. Added one remark. Apart from that looks good.

**Comment by kemingy** [comment]
BTW, this will need to create a nightly release in case to get the `git diff`. You can trigger the nightly action manually.

**Comment by mre** [comment]
Thanks! I've created the token now and added it. Let's see if it works. Triggering a build.

**Comment by mre** [comment]
https://github.com/lycheeverse/lychee/actions/runs/5842853562

**Comment by mre** [comment]
First build worked. Second build failed: https://github.com/lycheeverse/lychee/actions/runs/5842863869/job/15844411326

**Comment by kemingy** [comment]
Hi @mre, this nightly trigger has two runs:
1. Manually triggered workflow_dispatch: will create a release with the provided tag, but doesn't delete if the tag already exists (I created this one only for debugging and running the first time purpose)
2. Scheduled at 8:45 UTC: will check if there are any new commits comparing the previous nightly to the current HEAD. If so, will try to delete the previous nightly release and create a new one.

The second one will be mainly used for the nightly release. We don't need to manually trigger it anymore.

I'll keep an eye on the recent changes to see if the nightly release can work as expected.

**Comment by mre** [comment]
Thanks for the help with this.
Two more things I've wondered about:
* shall we mention the git commit somewhere in the release that the nightly is based on? Either in the release notes or the asset names?
* shall we create the binary assets as well? Right now it's just the tar archive of the source.

**Comment by kemingy** [comment]
>     * shall we mention the git commit somewhere in the release that the nightly is based on? Either in the release notes or the asset names?
> 
>     * shall we create the binary assets as well? Right now it's just the tar archive of the source.

1. Yes I add the `gh release create nightly --generate-notes` so it will generate the notes automatically like https://github.com/lycheeverse/lychee/releases/tag/nightly
2. Yes it will upload the binary assets, I think it's still in progress https://github.com/lycheeverse/lychee/actions/runs/5846357265

**Comment by mre** [comment]
Awesome work! Looks nice. 👌

**Pull request merged**